opt/scopeserver/dataserver/utils/search.py

```diff
@@ -54,26 +54,39 @@ class ResultTypePair(NamedTuple):
 }
 
 
-def match_result_cost(term: str, result: str) -> Optional[int]:
+def match_result_cost(term: str, result: str, result_type: str) -> Tuple[int, Optional[int]]:
+    term_cost: Optional[int]
+
     if term == result:
-        return 0
-    if term.casefold() == result.casefold():
-        return 1
-    if result.startswith(term) or result.endswith(term):
-        return 2
-    if result.casefold().startswith(term.casefold()) or result.casefold().endswith(term.casefold()):
-        return 3
-    if term in result:
-        return 4
-    if term.casefold() in result.casefold():
-        return 5
-    return None
+        term_cost = 0
+    elif term.casefold() == result.casefold():
+        term_cost = 1
+    elif result.startswith(term) or result.endswith(term):
+        term_cost = 2
+    elif result.casefold().startswith(term.casefold()) or result.casefold().endswith(term.casefold()):
+        term_cost = 3
+    elif term in result:
+        term_cost = 4
+    elif term.casefold() in result.casefold():
+        term_cost = 5
+    else:
+        term_cost = None
+
+    if result_type in ["gene", "regulon"] or result_type.startswith("Clustering:"):
+        type_cost = 0
+    else:
+        type_cost = 1
+
+    return (type_cost, term_cost)
 
 
 def sort_results(search_term: str, results: List[MatchResult]) -> List[MatchResult]:
-    costs = [(match_result_cost(search_term, match.element), match, result) for match, result in results]
+    costs = [
+        (match_result_cost(search_term, match.element, match.element_type), match, result) for match, result in results
+    ]
     sorted_costs = sorted(
-        [(cost, match, result) for cost, match, result in costs if cost is not None], key=lambda cost: cost[0]
+        [(cost, match, result) for cost, match, result in costs if cost[1] is not None],
+        key=lambda cost: (cost[0][0], cost[0][1]),
```

Comment:
This is where the complexity comes in. Lets avoid if possible.
```suggestion
        key=lambda cost: cost[0],
```

Could also use:
```python
import operator
operator.itemgetter(0)
```
rather than the lambda.